File: src/java/y2022/Day1.java
package y2022;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.logging.Level.INFO;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.logging.Logger;
import java.util.stream.Collectors;
import java.util.stream.Stream;


public class Day1 {

  private static final Logger LOGGER = Logger.getLogger(Day1.class.getName());

  public static void main(String[] args) throws IOException {
    final List<String> input;
    try (Stream<String> stream = Files.lines( Paths.get("input/2022/day1"), UTF_8)) {
      input = stream.collect(Collectors.toList());
    }

    int currentMax = 0;
    int current = 0;

    Iterator<String> iterator = input.iterator();

    List<Integer> backpacks = new ArrayList<>();

    while (iterator.hasNext()) {
      String next = iterator.next();

      if (next.isEmpty()) {
        if (current > currentMax) {
          currentMax = current;
        }

        backpacks.add(current);
        current = 0;
      } else {
        current += Integer.parseInt(next);
      }

    }

    LOGGER.log(INFO, "Part 1: {0}", Integer.toString(currentMax));

    int part2 = backpacks.stream()
    .sorted(Comparator.reverseOrder())
    .limit(3)
    .collect(Collectors.summingInt(Integer::intValue));
  
    LOGGER.log(INFO, "Part 2: {0}", part2);
  }
}

File: src/kotlin/y2021/Day1.kt
package y2021

import java.nio.file.Files
import java.nio.file.Paths


fun main() {
    val lines = Files.readAllLines(Paths.get("input/2021/day1"))

    imperative(lines)
    functional(lines)
}

fun functional(lines: List<String>) {
    val iterator = lines.iterator()

    fun checkIncreases(previous: Int): Int {
        val current = iterator.next().toInt()
        val isIncrease = if (current > previous) 1 else 0
        return isIncrease + if (iterator.hasNext()) checkIncreases(current) else 0
    }

    println("Part 1, functional: ${checkIncreases(Int.MAX_VALUE)}")

    val anotherIterator = lines.iterator()
    fun checkIncreasesSliding(previous: Int) {

    }
}



fun imperative(lines: List<String>) {
    var count = 0
    var current = Int.MAX_VALUE

    lines.forEach{
        val next = it.toInt()
        if (next > current) {
            count++
        }
        current = next
    }

    println("Part 1, imperative: $count")

    count = 0
    val numbers = lines.map{ it.toInt() }

    var currentWindow = numbers[0] + numbers[1] + numbers[2]
    var index = 3
    do {
        val nextWindow = currentWindow - numbers[index - 3] + numbers[index]
        if (nextWindow > currentWindow) {
            count++
        }
        currentWindow = nextWindow
        index++
    } while (index < numbers.size)

    println("Part 2, imperative: $count")
}

